runtime/recipe-index.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * @license
+ * Copyright (c) 2018 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Loader} from './loader.js';
+import {Manifest} from './manifest.js';
+import {Arc} from './arc.js';
+import {SlotComposer} from './slot-composer.js';
+import {Strategizer} from '../strategizer/strategizer.js';
+import {StrategyExplorerAdapter} from './debug/strategy-explorer-adapter.js';
+import {Tracing} from '../tracelib/trace.js';
+import {ConvertConstraintsToConnections} from './strategies/convert-constraints-to-connections.js';
+import {ResolveRecipe} from './strategies/resolve-recipe.js';
+import {InitPopulation} from './strategies/init-population.js';
+import {AddUseHandles} from './strategies/add-use-handles.js';
+import * as Rulesets from './strategies/rulesets.js';
+
+export class RecipeIndex {
+  constructor(arc) {
+    let trace = Tracing.start({cat: 'index', name: 'RecipeIndex::constructor', overview: true});
+    let arcStub = new Arc({
+      id: 'index-stub',
+      slotComposer: new SlotComposer({affordance: 'mock', noRootContext: true}),
+      recipeIndex: {
+        recipes: []
+      },
+    });
+    this._strategizer = new Strategizer(
+      [
+        new InitPopulation(arc),
+        ...[
+          ConvertConstraintsToConnections,
+          AddUseHandles,
+          ResolveRecipe,
+        ].map(s => new s(arcStub))
+      ],
+      [],
+      Rulesets.Empty
+    );
+    this._recipes = trace.endWith(new Promise(async resolve => {
```

Comment:
Instead of having `_recipes` be a promise, you could lift the async behavior to a static constructor eg. `await RecipeIndex.newRecipeIndex(...)` => `new RecipeIndex([...recipes])`

Reply:
Yes, but (as discussed offline) RecipeIndex is instantiated in the arc constructor, which is not async, so there would still be asynchrony. In this example I don't know how to get rid of it - maybe it will become clearer once we use the index more?